src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -91,5 +122,99 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
 
             return default;
         }
+
+        /// <summary>Returns the element at a specified index in a sequence or a default value if the index is out of range.</summary>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <param name="index">The index of the element to retrieve, which is either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.
+        /// </exception>
+        /// <returns>
+        ///   <see langword="default" /> if <paramref name="index" /> is outside the bounds of the <paramref name="source" /> sequence; otherwise, the element at the specified position in the <paramref name="source" /> sequence.
+        /// </returns>
+        public static TSource? ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (!index.IsFromEnd)
+            {
+                return source.ElementAtOrDefault(index.Value);
+            }
+
+            TryGetElementAt(source, index.Value, out TSource? result);
+            return result;
+        }
+
+        private static bool TryGetElementAt<TSource>(IEnumerable<TSource> source, int indexFromEnd, [MaybeNullWhen(false)] out TSource? result)
+        {
+            Debug.Assert(indexFromEnd >= 0);
+
+            result = default;
+            if (indexFromEnd == 0)
+            {
+                return false;
+            }
+
+            if (source is IPartition<TSource> partition)
+            {
+                int count = partition.GetCount(onlyIfCheap: true);
+                if (count == 0)
+                {
+                    return false;
+                }
+
+                if (count > 0)
+                {
+                    if (indexFromEnd <= count)
+                    {
+                        result = partition.TryGetElementAt(count - indexFromEnd, out bool found);
+                        return found;
+                    }
+
+                    return false;
+                }
+            }
+            else if (source is IList<TSource> list)
+            {
+                int count = list.Count;
+                if (indexFromEnd <= count)
+                {
+                    result = list[count - indexFromEnd];
+                    return true;
+                }
+
+                return false;
+            }
+
+            using IEnumerator<TSource> e = source.GetEnumerator();
```

Comment:
Many enumerable types in `System.Collection` implement `ICollection` without implementing `IList<T>`. Given that an `ICollection` implementation only requires _O(1)_ space, would it make sense to consider an optimization for that type as well? /cc @stephentoub

Reply:
> would it make sense to consider an optimization for that type as well?

That does what?

Reply:
Use `ICollection.Count` to transform a fromEnd index to its regular index. Assuming we do this at enumeration time then changing counts are not a concern?

Reply:
> Use ICollection.Count to transform a fromEnd index to its regular index. Assuming we do this at enumeration time then changing counts are not a concern?

Oh, you meant not to avoid enumerating (it would still need to do that), but rather to avoid employing a queue?  Sure.

Reply:
The queue is not needed for a type with `Count`. This can be a significant performance gain for large collection instances.

System.Linq reference to System.Runtime and System.Collection. There are 3 interfaces with the `Count` property:
- `ICollection`
- `ICollection<T>`
- `IReadOnlyCollection<T>`

@stephentoub @eiriktsarpalis Should I detect all 3 types? System.Linq does not detect `IReadOnlyCollection<T>` anywhere else.

As previously discussed, `Count` is not detected in `Take`, because `Take` has deferred execution and `source` can be mutable.

Reply:
I would stick with just `ICollection<T>`.  If we want to recognize `IReadOnlyCollection<T>`, that should be done across LINQ, and will need further discussion and performance measurement (there have been many issues about it over the years).

Reply:
> I would stick with just `ICollection<T>`.

@stephentoub How about `ICollection` the non-generic interface?

Reply:
Barring a couple of exceptions, we don't seem to test for non-generic `ICollection` in Linq. 

Related to https://github.com/dotnet/runtime/issues/28651.

Reply:
Updated the code to detect `ICollection<T>` and `ICollection`. Updated the existing APIs as well `ElementAt(int)` and `ElementAtOrDefault(int)`.

Reply:
Based on other Linq implementations I would contend that `ICollection` is probably not necessary. 

Also, do you think it might make sense to apply a similar optimization in the Take implementation?

Reply:
>apply a similar optimization in the Take implementation?

For "similar optimization", I assume you mean to detect `IList<T>` and `ICollection<T>`, get the `Count`, so that a queue is not needed.

I am not sure about this. `Take` has deferred execution. In other LINQ operators with deferred execution, I didn't see anyone detecting `IList<T>` and `ICollection<T>` and use the `Count`. This detection and `Count` is used in operators without deferred execution, like `ElementAt`, `ToArray`, etc.

Reply:
> I am not sure about this. Take has deferred execution. 

I think it should not be an issue as long `Count` is evaluated by the enumerator rather than at enumerable construction time. Essentially a type test could be done [at this point](https://github.com/dotnet/runtime/pull/47950/files#diff-8e4d6fffcdc28ad1a5d50050c7a22b13d183971a3f7d779df99f0088794b5ee7R74). A downside of that approach is that we won't be able take advantage of IPartitioner optimizations, which ties into my https://github.com/dotnet/runtime/pull/47950#discussion_r572776032 that we should perhaps author a general-purpose `Take(Range)` iterator that lives in speedopt/sizeopt.

> In other LINQ operators with deferred execution, I didn't see anyone detecting IList<T> and ICollection<T> and use the Count.

Avoiding _O(n)_ allocations is probably something that could motivate us in this case.

Reply:
> I think it should not be an issue as long Count is evaluated by the enumerator rather than at enumerable construction time.

Is this the final decision? If so I'll update the code.

If we are going in this direction, should `TakeLast` and `SkipLast` be updated to detect `Count` as well, to keep the consistency?

Another possible option is to update `IPartition<T>` to support indexes from end. Then use `IPartition<T>` to implement `Take(Range)`, `Take(int)`, `Skip(int)`, `TakeLast(int)`, `SkipLast(int)`. This should ultimately resolve the issues and keep everything consistent.

@eiriktsarpalis @stephentoub Please let me know what is the decision.

Reply:
My thinking was to use one ranged `TakeIterator` implementation (both for `speedopt` and `sizeopt`) that all of `Take(int)`, `Skip`, `TakeLast` and `SkipLast` can be based on. But perhaps it's outside of scope for the particular PR, I guess I'm fine with investigating that avenue at a later time.

Reply:
## ElementAt

For `ElementAt`, regarding the `count` detecting

> non-generic ICollection is probably not necessary.

I suggest detecting non-generic ICollection for consistency. I looked through other methods, like LINQ `Count()` operator and `IPartition.GetCount()` implementations, they detect `ICollection<T>`, non-generic `ICollection`, and `IIListProvider<T>`. I updated the code to do the same detections in the same order as `Count()`.

## Take

For `Take(Range)`, in the scope of this PR, I suggest keeping the current implementation, which is consistent with `TakeLast` and `SkipLast`. 

> My thinking was to use one ranged TakeIterator implementation (both for speedopt and sizeopt) that all of Take(int), Skip, TakeLast and SkipLast can be based on. But perhaps it's outside of scope for the particular PR, I guess I'm fine with investigating that avenue at a later time.

Agree. I also feel all the mentioned optimizations are more proper to go to the `IPartition<T>` level, where all existing optimizations are located. Once this PR gets in, I will to propose an optimization of partitioning (already working on the coding):

- Update `IPartition<T>` to support indexes from start and end, and update the implementations:
  - EmptyPartition
  - EnumerablePartition
  - ListPartition
  - RangeIterator
  - RepeatIterator
  - SelectArrayIterator
  - SelectIListIterator
  - SelectIPartitionIterator
  - SelectListIterator
  - SelectListPartitionIterator
  - SelectRangeIterator
  - OrderedPartition
  - OrderedEnumerable

- Implement `TakeIterator` based on `IPartition<T>`s. All the above-mentioned optimizations should go to `IPartition<T>`s, not in `TakeIterator`.

- Implement `Take(Range)`, `Take(int)`, `Skip(int)`, `TakeLast(int)`, `SkipLast(int)` using `TakeIterator`.

Then everything should be optimized, consolidated and consistent.

Reply:
## ElementAt

For `ElementAt`, regarding the `count` detecting

> non-generic ICollection is probably not necessary.

I suggest detecting non-generic ICollection for consistency. I looked through other methods, like LINQ `Count()` operator and `IPartition.GetCount()` implementations, they detect `ICollection<T>`, non-generic `ICollection`, and `IIListProvider<T>`. I updated the code to do the same detections in the same order as `Count()`.

## Take

For `Take(Range)`, in the scope of this PR, I suggest keeping the current implementation, which is consistent with `TakeLast` and `SkipLast`. 

> My thinking was to use one ranged TakeIterator implementation (both for speedopt and sizeopt) that all of Take(int), Skip, TakeLast and SkipLast can be based on. But perhaps it's outside of scope for the particular PR, I guess I'm fine with investigating that avenue at a later time.

Agree. I also feel all the mentioned optimizations are more proper to go to the `IPartition<T>` level, where all existing optimizations are located. Once this PR gets in, I will to propose an optimization of partitioning (already working on the coding):

- Update `IPartition<T>` to support indexes from start and end, and update the implementations:
  - EmptyPartition
  - EnumerablePartition
  - ListPartition
  - RangeIterator
  - RepeatIterator
  - SelectArrayIterator
  - SelectIListIterator
  - SelectIPartitionIterator
  - SelectListIterator
  - SelectListPartitionIterator
  - SelectRangeIterator
  - OrderedPartition
  - OrderedEnumerable

- Implement `TakeIterator` based on `IPartition<T>`s. All the above-mentioned optimizations should go to `IPartition<T>`s, not in `TakeIterator`.

- Implement `Take(Range)`, `Take(int)`, `Skip(int)`, `TakeLast(int)`, `SkipLast(int)` using `TakeIterator`.

Then everything should be optimized, consolidated and consistent.

@eiriktsarpalis @stephentoub Please let me know what you think.

Reply:
Agreed.